---
title: "R Notebook"
output:
  pdf_document: default
  html_notebook: default
---

```{r}
library(GEOquery)
library(Biobase)
library(DESeq2)
library(RColorBrewer)
library(pheatmap)
library(tidyverse)
#library(apeglm)
library(AnnotationDbi)
library(org.Mm.eg.db)
library(limma)
```

## Download soft files from GEO

```{r}
gse <- getGEO(filename = "./data/GSE48962/GSE48963_family.soft")
```

```{r}
samples_info <- data.frame(matrix(ncol = 4, nrow = 0))
colnames(samples_info) <- c("geo_accession", "title", "library_strategy", "source_name_ch1")

for(gsm in GSMList(gse)) {
  gsm_meta <- Meta(gsm)
  samples_info <- samples_info %>%
    add_row(
      geo_accession    = gsm_meta$geo_accession,
      title            = gsm_meta$title,
      library_strategy = gsm_meta$library_strategy,
      source_name_ch1  = gsm_meta$source_name_ch1
    )
}
```

## Filter samples
Select RNA-seq samples, 12 weeks
.
```{r}
samples_rna_12 <- samples_info %>%
  separate(source_name_ch1, c("genotype", "tissue", "age"), ", ") %>%
  mutate(genotype = ifelse(genotype == "R6/2", "HD", "WT")) %>%
  filter(library_strategy == "RNA-Seq", age == "12 weeks", tissue == "cortex") %>%
  dplyr::select(-c(library_strategy, age))
samples_rna_12
```

## Prepare data

```{r}
SAMPLES_FOLDER = "./data/GSE48962/"

read_counts <- function(sample_info) {
  file_path <- paste(SAMPLES_FOLDER, sample_info$geo_accession, "_", sample_info$title, ".combined.rpkm", sep = "")

  return (
    read.table(file_path, sep = "\t", row.names = 1) %>%
      dplyr::select(V2) %>%
      dplyr::rename(!!sample_info$title := V2)
  )
}

total_counts <- read_counts(samples_rna_12[1, ])

for(i in 2:nrow(samples_rna_12)) {
  sample_counts <- read_counts(samples_rna_12[i, ])
  
  total_counts <- merge(total_counts, sample_counts, by = 0, all = TRUE)
  rownames(total_counts) = total_counts[, "Row.names"]
  total_counts <- total_counts %>% dplyr::select(-Row.names)
}
```

```{r}
head(total_counts)
```

Check for duplicates:

```{r}
sum(duplicated(rownames(total_counts)))
```


## A note on counts

We do not perform any filtering based on minium read count, since it only affects the computation speed and object size. Such transformations are not providing any benefits for a dataset of 12 samples.

## DESeq2

Create columns metadata for DESeq2:

```{r}
col_data <- samples_rna_12 %>%
  mutate(group = paste(genotype, tissue, sep = "_")) %>%
  magrittr::set_rownames(.$title)
head(col_data)
```


```{r}
dds <- DESeqDataSetFromMatrix(countData = total_counts,
                                    colData = col_data,
                                    design = ~ group)
```

## PCA

```{r}
dds <- estimateSizeFactors(dds)
normalized_counts <- counts(dds, normalized = TRUE)
vsd <- vst(dds, blind = TRUE)

plotPCA(vsd, intgroup = "group") +
  geom_label(aes(label = name))
# percentVar <- round(100 * attr(pcaData, "percentVar"))
# ggplot(pcaData, aes(PC1, PC2, color=group)) +
#   geom_point(size=3) +
#   xlab(paste0("PC1: ",percentVar[1],"% variance")) +
#   ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
#   geom_label(aes(label = name)) +
#   coord_fixed()
```
All the groups can be separated based on PC1 and PC2.


## Dispersion plot

```{r}
DESeq2::plotDispEsts(de_result)
```

## Differential Expression

Peform analysis:

```{r}
de_result <- DESeq(dds)
```

Define contrasts for comparisons within tissue:

```{r}
contrasts_striatum <- c("group", "HD_striatum", "WT_striatum")
contrasts_cortex <- c("group", "WT_cortex", "HD_cortex")
```

```{r}
res_striatum <- results(de_result, lfcThreshold = 0.5, contrast = contrasts_striatum) %>%
  as.data.frame() %>%
  drop_na()
```

## MA Plot

```{r}
res_striatum %>%
  mutate(significant = padj < 0.1) %>%
  dplyr::select(baseMean, log2FoldChange, significant) %>%
  plotMA()

abline(h=c(-0.5, 0.5), col="dodgerblue", lwd=2)
```

## MA Plot(shrinkage)

```{r}
# res_cortex_shrink <- lfcShrink(de_result, coef="group_WT_cortex_vs_HD_cortex", type="apeglm")
#   #arrange(baseMean) %>%
#   #select(baseMean, log2FoldChange, padj < 0.1) %>%
# summary(res_cortex_shrink)
# plotMA(res_cortex_shrink)
```

## Select significant results

Add gene symbol:

```{r}
entrez_id <- rownames(res_striatum)
res_striatum$gene_symbol <- mapIds(org.Mm.eg.db,
                     keys = entrez_id,
                     column = "SYMBOL",
                     keytype = "ENTREZID",
                     multiVals = "first")
```

```{r}
res_striatum_sig <- res_striatum %>%
  filter(padj < 0.1) %>%
  arrange(padj) 
```

```{r}
res_striatum_sig %>%
  write_delim('./out/GSE48963_str_12_sig.tsv')
```

## Export for BETA analysis

```{r}
res_striatum %>%
  dplyr::select(gene_symbol, log2FoldChange, padj) %>%
  write_delim('./out/GSE48963_str_12_all.tsv', delim = "\t", col_names = F)
```


## Cortext

```{r}
res_cortex <- results(de_result, lfcThreshold = 0.5, altHypothesis = "greaterAbs", contrast = contrasts_cortex) %>%
  as.data.frame() %>%
  drop_na()
```

```{r}
res_cortex %>%
  mutate(significant = padj < 0.1) %>%
  dplyr::select(baseMean, log2FoldChange, significant) %>%
  plotMA()

abline(h=c(-0.5, 0.5), col="dodgerblue", lwd=2)
```

```{r}
entrez_id <- rownames(res_cortex)
res_cortex$gene_symbol <- mapIds(org.Mm.eg.db,
                     keys = entrez_id,
                     column = "SYMBOL",
                     keytype = "ENTREZID",
                     multiVals = "first")
```

```{r}
res_cortex_sig <- res_cortex %>%
  filter(padj < 0.1) %>%
  arrange(padj) 
```

```{r}
res_cortex %>%
  dplyr::select(gene_symbol, log2FoldChange, padj) %>%
  write_delim('./out/GSE48963_ctx_12_all.tsv', delim = "\t", col_names = F)
```

```{r}
res_cortex_sig %>%
  filter(padj < 0.1) %>%
  dplyr::select(gene_symbol) %>%
  write_delim('./out/GSE48963_ctx_12_sig_symbols.tsv', delim = "\t", col_names = F)
```









