---
title: "R Notebook"
output:
  pdf_document: default
  html_notebook: default
---

```{r}
library(GEOquery)
library(Biobase)
library(DESeq2)
library(RColorBrewer)
library(pheatmap)
library(tidyverse)
library(AnnotationDbi)
library(org.Mm.eg.db)
library(limma)
```

## Download soft files from GEO

```{r}
gse <- getGEO(filename = "./data/GSE48962/GSE48963_family.soft")
```

```{r}
samples_info <- data.frame(matrix(ncol = 4, nrow = 0))
colnames(samples_info) <- c("geo_accession", "title", "library_strategy", "source_name_ch1")

for(gsm in GSMList(gse)) {
  gsm_meta <- Meta(gsm)
  samples_info <- samples_info %>%
    add_row(
      geo_accession    = gsm_meta$geo_accession,
      title            = gsm_meta$title,
      library_strategy = gsm_meta$library_strategy,
      source_name_ch1  = gsm_meta$source_name_ch1
    )
}
```

## Filter samples
Select RNA-seq samples, 12 weeks
.
```{r}
samples_rna_12 <- samples_info %>%
  separate(source_name_ch1, c("genotype", "tissue", "age"), ", ") %>%
  mutate(genotype = ifelse(genotype == "R6/2", "HD", "WT")) %>%
  filter(library_strategy == "RNA-Seq", age == "12 weeks") %>%
  dplyr::select(-c(library_strategy, age))
samples_rna_12
```

## Prepare data

```{r}
SAMPLES_FOLDER = "./data/GSE48962/"

read_counts <- function(sample_info) {
  file_path <- paste(SAMPLES_FOLDER, sample_info$geo_accession, "_", sample_info$title, ".combined.rpkm", sep = "")

  return (
    read.table(file_path, sep = "\t", row.names = 1) %>%
      dplyr::select(V2) %>%
      dplyr::rename(!!sample_info$title := V2)
  )
}

total_counts <- read_counts(samples_rna_12[1, ])

for(i in 2:nrow(samples_rna_12)) {
  sample_counts <- read_counts(samples_rna_12[i, ])
  
  total_counts <- merge(total_counts, sample_counts, by = 0, all = TRUE)
  rownames(total_counts) = total_counts[, "Row.names"]
  total_counts <- total_counts %>% dplyr::select(-Row.names)
}
```

```{r}
head(total_counts)
```

## A note on counts

We do not perform any filtering based on minium read count, since it only affects the computation speed and object size. Such transformations do not providing any benefits for a small dataset.

## DESeq2

Create columns metadata for DESeq2:

```{r}
col_data <- samples_rna_12 %>%
  mutate(group = paste(genotype, tissue, sep = "_")) %>%
  magrittr::set_rownames(.$title)
col_data
```

### PCA for all groups

```{r}
dds_all <- DESeqDataSetFromMatrix(countData = total_counts,
                                    colData = col_data,
                                    design = ~ group)
dds_all <- estimateSizeFactors(dds_all)
normalized_counts <- counts(dds_all, normalized = TRUE)
vsd_all <- vst(dds_all, blind = TRUE)

group_labels <- c("HD, cortex", "HD, striatum", "WT, cortex", "WT, striatum")
group_colors <- c("#CC79A7", "#E69F00", "#56B4E9", "#009E73")
plotPCA(vsd_all, intgroup = "group") +
  scale_color_manual(labels = group_labels, values = group_colors) +
  theme_minimal() +
  coord_fixed(ratio = sqrt(20 / 65))
```
### Helper functions

```{r}
plotMA_with_line <- function(res) {
  res %>%
    mutate(significant = padj < 0.1) %>%
    dplyr::select(baseMean, log2FoldChange, significant) %>%
    plotMA()
    abline(h=c(-0.5, 0.5), col="dodgerblue", lwd = 2)
}

get_results <- function(res, contrasts) {
  return (
    results(res, lfcThreshold = 0.5, contrast = contrasts) %>%
      as.data.frame() %>%
      drop_na()
  )
}

add_gene_symbols <-function(res) {
  res$gene_symbol <- mapIds(org.Mm.eg.db,
                       keys = rownames(res),
                       column = "SYMBOL",
                       keytype = "ENTREZID",
                       multiVals = "first")
  res$entrez_id <- rownames(res)
  
  return(res)
}

select_sig <- function(res) {
  return (
    res %>%
      filter(padj < 0.1) %>%
      arrange(padj)
  )
}

save_result <- function(res, out_file) {
  res %>%
    dplyr::select(entrez_id, gene_symbol, log2FoldChange, padj) %>%
    write_delim(out_file, delim = "\t", col_names = T)
}
```

###  Differential Expression in Cortex

```{r}
col_data_ctx <- col_data %>%
  filter(tissue == "cortex")
counts_ctx <- total_counts[, col_data_ctx %>% dplyr::select(title) %>% unlist()]

dds_ctx <- DESeqDataSetFromMatrix(countData = counts_ctx,
                                    colData = col_data_ctx,
                                    design = ~ group)
de_result_ctx <- DESeq(dds_ctx)
contrasts_ctx <- c("group", "WT_cortex", "HD_cortex")
res_ctx <- get_results(de_result_ctx, contrasts_ctx)
DESeq2::plotDispEsts(de_result_ctx)
plotMA_with_line(res_ctx)
```

```{r}
res_ctx <- add_gene_symbols(res_ctx)
res_ctx_sig <- select_sig(res_ctx)
save_result(res_ctx_sig, './out/GSE48963_ctx_12_sig.tsv')
```

## Differential Expression in Striatum

```{r}
col_data_str <- col_data %>%
  filter(tissue == "striatum")
counts_str <- total_counts[, col_data_str %>% dplyr::select(title) %>% unlist()]

dds_str <- DESeqDataSetFromMatrix(countData = counts_str,
                                    colData = col_data_str,
                                    design = ~ group)
de_result_str <- DESeq(dds_str)
contrasts_str <- c("group", "HD_striatum", "WT_striatum")
res_str <- get_results(de_result_str, contrasts_str)
DESeq2::plotDispEsts(de_result_str)
plotMA_with_line(res_str)
```


```{r}
res_str <- add_gene_symbols(res_str)
res_str_sig <- select_sig(res_str)
save_result(res_str_sig, './out/GSE48963_str_12_sig.tsv')
```









